### PR TITLE
[Inactive] Changes for basic LLaDA style diffusion masking support

### PR DESCRIPTION
**Pull request by gopeshh**
# ✨ Description

Cleaned up the code a bit:
1) Added Diffusion config object as we discussed
2) removed noise schedules for v1
3) Moved loss calculation to head.py (as I noticed language modelling loss is computed there)
4) Moved bidirectional attention to preprocessing.py file as it seems like the attention mask is computed there

Of course still a WIP but feel free to leave comments and suggestions

These are changes to address this PR: https://github.com/ServiceNow/Fast-LLM/issues/208#issue-2950083282

### CODE REVIEW

File: fast_llm/layers/transformer/preprocessing.py [hunk 1]
```diff
@@ -337,3 +338,75 @@ def preprocess(self, batch, kwargs: dict[str, typing.Any]) -> None:
...
+        
+        kwargs['masked_indices'] = masked_indices
+        kwargs['p_mask'] = p_mask
+        
+        if self._config.diffusion.bidirectional_attention:
+            # Bidirectional attention - all tokens can attend to all other tokens
+            attention_mask = torch.ones((batch_size, 1, seq_len, seq_len), device=device, dtype=torch.bool)
+        else:
+            # Causal attention
+            attention_mask = torch.ones((batch_size, 1, seq_len, seq_len), device=device, dtype=torch.bool).tril_()
```
[the remark this answers is not in the document]
Comment (gopeshh) [reply fast_llm/layers/transformer/preprocessing.py:384]: Yes, permuted rows and columns makes sense - so we can preserve the order in which it was unmasked. I will update this.
I guess this idea is similar to this paper? https://arxiv.org/abs/1906.08237

File: fast_llm/layers/transformer/preprocessing.py [hunk 2]
```diff
@@ -337,3 +338,75 @@ def preprocess(self, batch, kwargs: dict[str, typing.Any]) -> None:
...
+        
+        if diffusion_config.pad_prob > 0:
+            pad_mask = torch.rand((batch_size,), device=device) < diffusion_config.pad_prob
+            if pad_mask.any():
+                masked_indices[pad_mask] = True
+        
+        kwargs['masked_indices'] = masked_indices
+        kwargs['p_mask'] = p_mask
+        
+        if self._config.diffusion.bidirectional_attention:
```
[the remark this answers is not in the document]
Comment (gopeshh) [reply fast_llm/layers/transformer/preprocessing.py:379]: I agree, will change this!

File: fast_llm/layers/transformer/preprocessing.py [hunk 3]
```diff
@@ -337,3 +338,75 @@ def preprocess(self, batch, kwargs: dict[str, typing.Any]) -> None:
...
+        diffusion_config = kwargs['parameters'].diffusion
+        if not diffusion_config.enabled:
+            return
+            
+        batch_size, seq_len = batch.shape
+        device = batch.device
+        
+        t = torch.rand(batch_size, device=device)
+        
+        p_mask = (1 - diffusion_config.epsilon) * t + diffusion_config.epsilon
```
[the remark this answers is not in the document]
Comment (gopeshh) [reply fast_llm/layers/transformer/preprocessing.py:365]: Are you saying this coz we could have many timesteps with the exact masking level set to `max_mask_prob`? So are you suggesting some soft clipping instead of a hard upper bound?

File: fast_llm/layers/transformer/preprocessing.py [hunk 4]
```diff
@@ -337,3 +338,75 @@ def preprocess(self, batch, kwargs: dict[str, typing.Any]) -> None:
...
+        p_mask = torch.min(p_mask, torch.tensor(diffusion_config.max_mask_prob))
+        p_mask = p_mask[:, None].expand(-1, seq_len)
+        
+        masked_indices = torch.rand((batch_size, seq_len), device=device) < p_mask
+        
+        if diffusion_config.pad_prob > 0:
```
[the remark this answers is not in the document]
Comment (nitsanluke) [reply fast_llm/layers/transformer/preprocessing.py:371]: Is this to include variable length sequences for 1% of the data?

[the remark this answers is not in the document]
Comment (gopeshh) [reply fast_llm/layers/transformer/preprocessing.py:371]: Yeah exactly!

**Comment by jlamypoirier** [comment]
Is this still relevant?
